### PR TITLE
Add CancellationTokenSource.TryReset

### LINKED ISSUE #48492: Add CancellationTokenSource.TryReset()
## Background and Motivation

When a library or framework exposes a `CancellationToken` that usually does not get canceled (e.g. [HttpContext.RequestAborted](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.httpcontext.requestaborted?view=aspnetcore-5.0)) they often still need to dispose the backing `CancellationTokenSource` (CTS) after an operation completes rather than reusing it in order to account for callers that might never dispose their registrations.

To use the `RequestAborted` example, Kestrel disposes the backing CTS after every request where the `RequestAborted` token is accessed. If Kestrel attempted to reuse the CTS backing the `RequestAborted` token for future requests in order to reduce allocations, it would risk leaking any undisposed registrations and triggering the undisposed registrations when unrelated requests are aborted.

Another scenario where people want to be able to "reset" a CTS is after calling `CancelAfter()`. This can already be achieved by calling `CancelAfter(Timeout.Infinite)`, but that's not necessarily obvious unless you read the docs. `TryReset()` is something that would immediately make sense in this scenario when looking at intellisense completions.

Another benefit is that it's immediately obvious if resetting failed. If you try resetting a timeout with `CancelAfter()`, you have to check after the call to verify their was no cancellation before or during the call causing `CancelAfter()` to no-op. This is demonstrated in the second `HttpClient` usage example.

## Proposed API

```diff
namespace System.Threading
{
    public class CancellationTokenSource : IDisposable
    {
+        // Returns false if the CTS has already been canceled
+        public bool TryReset();    
    }
}
```

## Usage Examples

Simplified `RequestAborted` example.

```csharp
private readonly CancellationTokenSource _abortCts = new CancellationTokenSource();

// Imagine this could be called in the background because a client disconnect or some timeout.
void AbortConnection()
{
    _abortCts.Cancel();
}

async Task ProcessRequestsAsync()
{
   while (_abortCts.TryReset())
   {
      var httpContext = await ParseRequest(_abortCts.Token);
      httpContext.RequestAborted = _abortCts.Token;
      await Middleware.InvokeAsync(httpContext);
   }
}
```

HttpClient request

```csharp
var cts = new CancellationTokenSource();
var httpClient = new HttpClient();

while (true)
{
    cts.CancelAfter(TimeSpan.FromSeconds(10));
    var response = await httpClient.GetAsync("http://example.org/healthcheck", cts.Token);
    response.EnsureSuccessStatusCode();
    
    if (!cts.TryReset())
    {
        cts = new CancellationTokenSource();
    }
}
```

Today, without `TryReset()`, you could achieve similar functionality like so:

```csharp
var cts = new CancellationTokenSource();
var httpClient = new HttpClient();

while (true)
{
    cts.CancelAfter(TimeSpan.FromSeconds(10));

    if (cts.Token.IsCancellationRequested)
    {
        cts = new CancellationTokenSource(TimeSpan.FromSeconds(10));
    }

    // ...

```

So in the case of timeouts, the proposed reset APIs don't offer any entirely new functionality, but it exposes it in a much more obvious way.

## Alternative Designs

Since calling `CancelAfter(Timeout.Infinite)` should already reset any timeouts, we could consider an API that just clears registrations. You could call both `CancelAfter(Timeout.Infinite)` and `TryClearRegistrations()` to get the full `TryReset()` behavior mentioned above.

```diff
namespace System.Threading
{
    public class CancellationTokenSource : IDisposable {
+        public bool TryClearRegistrations();    
    }
}
```

## Risks

People might assume they can reset a CTS that has already been canceled. We should clearly document this isn't the case. Hopefully the `Try` in the name will get people thinking about the cases in which it won't work.

Code that references a stale token and registers after the backing CTS has been reset can run into the same problems those who don't dispose their registrations today face when the CTS is reused. That might not sound bad, but this API will likely encourage more reuse of CTS objects. I do think that not disposing registrations is more common than using stale tokens though.

### PR DESCRIPTION
**Pull request by stephentoub**
Fixes https://github.com/dotnet/runtime/issues/48492

### CODE REVIEW

**Comment by GrabYourPitchforks** [comment]
Trying to wrap my mind around what happens if we have a linked CTS (via `CTS.CreateLinkedTokenSource`). Let's say that we have two parent CT instances: _parent1_ and _parent2_, and we create a child CTS _child_:

```cs
CancellationToken parent1 = GetToken();
CancellationToken parent2 = GetToken();
CancellationTokenSource child = CancellationTokenSource.CreateLinkedTokenSource(parent1, parent2);
```

From reading through this code, it looks like if either _parent1_'s CTS or _parent2_'s CTS is reset, then that particular token tripping will no longer trigger the child being canceled, but the other token can still trigger a cancellation. That seems fine.

But what happens if somebody calls _child.TryReset()_? It clears any registrations tied directly to it, but it's still listening for cancellation notifications coming from its parents, correct?

What I'm trying to figure out is how the phrase "usage of TryReset concurrently with issuing cancellation is not thread-safe" squares with the fact that we might have a linked CTS. Since I assume most linked CTS instances don't "own" the parent and can't control when the parent triggers cancellation, it would be ill-advised to call _TryReset_ on a CTS returned by the _CreateLinkedTokenSource_ API?

File: src/libraries/System.Private.CoreLib/src/System/Threading/CancellationTokenSource.cs [hunk 1]
```diff
@@ -365,6 +365,53 @@ private void CancelAfter(uint millisecondsDelay)
...
+        /// </summary>
+        /// <returns>
+        /// true if the <see cref="CancellationTokenSource"/> has not had cancellation requested and could
+        /// have its state reset to be reused for a subsequent operation; otherwise, false.
+        /// </returns>
+        /// <remarks>
+        /// <see cref="TryReset"/> is intended to be used by the sole owner of the <see cref="CancellationTokenSource"/>
+        /// when it is known that the operation with which the <see cref="CancellationTokenSource"/> was used has
+        /// completed, no one else will be attempting to cancel it (usage of <see cref="TryReset"/> concurrently with
+        /// issuing cancellation is not thread-safe), and any registrations still remaining are erroneous.  Upon a
```
Comment (davidfowl) [inline src/libraries/System.Private.CoreLib/src/System/Threading/CancellationTokenSource.cs:379]: > (usage of <see cref="TryReset"/> concurrently with issuing cancellation is not thread-safe)

Turns out this is fine for kestrel, we're already doing this today.

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Threading/CancellationTokenSource.cs:379]: > Turns out this is fine for kestrel, we're already doing this today.

TryReset doesn't exist today :)  What is it Kestrel is already doing?

**Comment by stephentoub** [comment]
> From reading through this code, it looks like if either parent1's CTS or parent2's CTS is reset, then that particular token tripping will no longer trigger the child being canceled, but the other token can still trigger a cancellation. That seems fine.

Correct.  If you have:
```C#
var cts2 = CreateLinkedTokenSource(cts1.Token);
```
then this is basically the same as:
```C#
var cts2 = new CancellationTokenSource();
cts2._registrationToDispose = cts1.Token.Register(() => cts2.Cancel());
```
So if you call cts1.TryReset, it'll drop the registration to call cts2.Cancel, and if cts1 has cancellation requested, it won't forward that along to cts2.

> But what happens if somebody calls child.TryReset()? It clears any registrations tied directly to it, but it's still listening for cancellation notifications coming from its parents, correct?

Correct

> What I'm trying to figure out is how the phrase "usage of TryReset concurrently with issuing cancellation is not thread-safe" squares with the fact that we might have a linked CTS. Since I assume most linked CTS instances don't "own" the parent and can't control when the parent triggers cancellation, it would be ill-advised to call TryReset on a CTS returned by the CreateLinkedTokenSource API?

Right.  It's no different than if you handed out a reference to your CTS to someone else that might call Cancel{After}.  You're no longer the sole owner, and so using TryReset is problematic.  You won't corrupt the instance, but you might return true even if the CTS is already canceled.

File: src/libraries/System.Private.CoreLib/src/System/Threading/CancellationTokenSource.cs (hunk 1 above)
Comment (davidfowl) [reply src/libraries/System.Private.CoreLib/src/System/Threading/CancellationTokenSource.cs:379]: Synchronizes throwing away the token and it potentially firing.

**Comment by GrabYourPitchforks** [comment]
> You won't corrupt the instance, but you might return true even if the CTS is already canceled.

That seems fine then. A race condition is much preferred to corruption. :)

**Review by GrabYourPitchforks — APPROVED** [review]

**Comment by stephentoub** [comment]
> That seems fine then. A race condition is much preferred to corruption. :)

I'll augment the comment to clarify. Thanks.

File: src/libraries/System.Private.CoreLib/src/System/Threading/CancellationTokenSource.cs [hunk 2]
```diff
@@ -365,6 +365,53 @@ private void CancelAfter(uint millisecondsDelay)
...
+        /// successful reset, such registrations will no longer be notified for any subsequent cancellation of the
+        /// <see cref="CancellationTokenSource"/>; however, if any component still holds a reference to this
+        /// <see cref="CancellationTokenSource"/> either directly or indirectly via a <see cref="CancellationToken"/>
+        /// handed out from it, polling via their reference will show the current state any time after the reset as
+        /// it's the same instance.
+        /// </remarks>
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Threading/CancellationTokenSource.cs:385]:
```suggestion
        /// <see cref="TryReset"/> is intended to be used by the sole owner of the <see cref="CancellationTokenSource"/>
        /// when it is known that the operation with which the <see cref="CancellationTokenSource"/> was used has
        /// completed, no one else will be attempting to cancel it, and any registrations still remaining are erroneous.
        /// Upon a successful reset, such registrations will no longer be notified for any subsequent cancellation of the
        /// <see cref="CancellationTokenSource"/>; however, if any component still holds a reference to this
        /// <see cref="CancellationTokenSource"/> either directly or indirectly via a <see cref="CancellationToken"/>
        /// handed out from it, polling via their reference will show the current state any time after the reset as
        /// it's the same instance.  Usage of <see cref="TryReset"/> concurrently with requesting cancellation is not
        /// thread-safe and may result in TryReset returning true even if cancellation was already requested and may result
        /// in registrations not being invoked as part of the concurrent cancellation request.
        /// </remarks>
```

**Comment by davidfowl** [comment]
It's time ❗ 

cc @BrennanConroy

**Pull request merged**